scripts-available/CDB_TableMetadata.sql

```diff
@@ -223,15 +214,39 @@ CREATE OR REPLACE FUNCTION public.CDB_TableMetadataTouch(tablename regclass)
     RETURNS void AS
     $$
     BEGIN
-        WITH upsert AS (
-            UPDATE public.cdb_tablemetadata
-            SET updated_at = NOW()
-            WHERE tabname = tablename
-            RETURNING *
+        WITH nv as (
+          SELECT tablename as tabname, NOW() as t
+        ), all_dependents as (
+          SELECT nv.tabname, nv.t
+          FROM nv
+          UNION
+          SELECT
+            dv.dependent_oid as tabname,
+            nv.t
+          FROM nv, public.CDB_TableMetadata_DependentViews(ARRAY[nv.tabname::regclass]) dv
+        ), updated as (
+          UPDATE public.CDB_TableMetadata x SET updated_at = ad.t
+          FROM (
+            SELECT
+              md.tabname,
+              ad.t
+            FROM public.CDB_TableMetadata md
+            JOIN all_dependents ad
+              ON ad.tabname = md.tabname
+            ORDER BY md.tabname ASC
+            FOR UPDATE OF md
+          ) ad
+          WHERE x.tabname = ad.tabname
+          RETURNING x.tabname
         )
-        INSERT INTO public.cdb_tablemetadata (tabname, updated_at)
-            SELECT tablename, NOW()
-            WHERE NOT EXISTS (SELECT * FROM upsert);
+        INSERT INTO public.CDB_TableMetadata SELECT ad.*
+        FROM all_dependents ad
+        LEFT JOIN updated u
+          ON u.tabname = ad.tabname
+        LEFT JOIN public.CDB_TableMetadata i
+          ON i.tabname = ad.tabname
+        WHERE u.tabname IS NULL
+          AND i.tabname IS NULL;
```

Comment:
I'm wondering what an explain of this query looks like, and if it would suffer under load (a lot of JOIN, albeit left joins). I'm not saying I don't like the implementation, I'm just pointing out we should stress-test this.

Reply:
The planner uses a hash anti join between two very small CTEs and can either use a seq scan or index seek on the primary key of `CDB_TableMetadata`, depending on which it determines is fastest based on table size/stats.  However, for even a table with 10 dependent views, a case which will likely never exist in practice, this takes only ~10ms.